{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Frequently Asked Questions\n",
    "\n",
    "This notebook has tips on how to access specific tables or queries from the generated tables. If you have any other question, please open an issue. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading the data\n",
    "Unless stated otherwise, the following function loads all the required data. The folder ```TablesForExploration``` can be found [here](https://opendata.earlham.ac.uk/wheat/under_license/toronto/Ramirez-Gonzalez_etal_2018-06025-Transcriptome-Landscape/data/TablesForExploration/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(gsubfn.engine = \"R\")\n",
    "library(ggplot2)\n",
    "library(reshape2)\n",
    "library(sqldf)\n",
    "library(fields)\n",
    "library(gridExtra)\n",
    "library(ggtern)\n",
    "library(clue)\n",
    "library(geometry)\n",
    "library(gtable)\n",
    "#library(goseq)\n",
    "library(plyr)\n",
    "library(scales)\n",
    "\n",
    "loadGeneInformation<-function(dir=\"../../TablesForExploration\", \n",
    "                              motifs=T, \n",
    "                              WGCNA=F, \n",
    "                              meanTpms=T\n",
    "                             ){\n",
    "    \n",
    "    path<-paste0(dir,\"/CanonicalTranscript.rds\")\n",
    "    canonicalTranscripts<-readRDS(path)\n",
    "    canonicalTranscripts$intron_length<- canonicalTranscripts$mrna_length -  canonicalTranscripts$exon_length\n",
    "    canonicalTranscripts$chr_group <- substr(canonicalTranscripts$Chr,4,4)\n",
    "    canonicalTranscripts$genome    <- substr(canonicalTranscripts$Chr,5,5)\n",
    "    expressed_genes <- canonicalTranscripts$Gene\n",
    "    \n",
    "    if(meanTpms == T){\n",
    "        path<-paste0(dir, \"/MeanTpms.rds\")\n",
    "        meanTpms <- readRDS(path)\n",
    "        expressed_genes<-unique(meanTpms$gene)\n",
    "    }\n",
    "    \n",
    "    canonicalTranscripts<-canonicalTranscripts[canonicalTranscripts$Gene %in% expressed_genes, ]\n",
    "    \n",
    "    canonicalTranscripts$scaled_5per_position <-   5 * ceiling(canonicalTranscripts$scaled_1per_position / 5)\n",
    "    canonicalTranscripts$scaled_5per_position <- ifelse(canonicalTranscripts$scaled_5per_position == 0, \n",
    "        5, \n",
    "        canonicalTranscripts$scaled_5per_position)\n",
    "\n",
    "    path<-paste0(dir, \"/region_partition.csv\")\n",
    "    partition<-read.csv(path, row.names=1)\n",
    "    \n",
    "    partition_percentages<-round(100*partition/partition$Length)\n",
    "    partition_percentages$Chr <- rownames(partition_percentages)\n",
    "    partition$Chr <- rownames(partition)\n",
    "    ct<-canonicalTranscripts\n",
    "    ct_with_partition<-sqldf('SELECT ct.*, CASE \n",
    "WHEN scaled_1per_position < R1_R2a THEN \"R1\"\n",
    "WHEN scaled_1per_position < R2a_C  THEN \"R2A\"\n",
    "WHEN scaled_1per_position < C_R2b  THEN \"C\"\n",
    "WHEN scaled_1per_position < R2b_R3  THEN \"R2B\"\n",
    "ELSE \"R3\" END as partition\n",
    "    \n",
    "FROM ct LEFT JOIN partition_percentages ON ct.chr = partition_percentages.chr   ')\n",
    "\n",
    "    x<-  as.factor(ct_with_partition$partition)\n",
    "    x <- factor(x,levels(x)[c(2,3,1,4,5)])\n",
    "    ct_with_partition$partition <- x \n",
    "    canonicalTranscripts<-ct_with_partition\n",
    "\n",
    "    path<-paste0(dir,\"/TriadMovement.rds\")\n",
    "    triadMovement<-readRDS(path)\n",
    "    \n",
    "    path<-paste0(dir,\"/Triads.rds\")\n",
    "    triads<-readRDS(path)\n",
    "    \n",
    "    path<-paste0(dir,\"/universe_table.csv\")\n",
    "    gene_universe<-read.csv(path)\n",
    "    \n",
    "    path<-paste0(dir, \"/OntologiesForGenes.rds\")\n",
    "    ontologies<-readRDS(path)\n",
    "    \n",
    "    path<-paste0(dir, \"/id_names_merged.txt\")\n",
    "    id_names <- read.csv(path, header=F, sep = \"\\t\")\n",
    "    \n",
    "    if(WGCNA == T){\n",
    "        path<-paste0(dir, \"/WGCNA_table.csv\")\n",
    "        WGCNA <-  read.csv(path)    \n",
    "    }\n",
    "    \n",
    "    path<-paste0(dir, \"/ObservedGOTermsWithSlim.csv\")\n",
    "    go_slim<-read.csv(path, row.names=1)\n",
    "\n",
    "    \n",
    "    if(motifs == T){\n",
    "        path<-paste0(dir, \"/motifs.rds\")\n",
    "        motifs <- readRDS(path)\n",
    "        motifs<-unique(motifs)\n",
    "    }\n",
    "   \n",
    "    path<-paste0(dir, \"/SegmentalTriads.csv\")\n",
    "    allTriads<-read.csv(path, stringsAsFactors=F)\n",
    "    only_genes<-allTriads[,c(\"group_id\",\"A\", \"B\", \"D\")]\n",
    "    allTriads<-melt(only_genes, id.vars<-c(\"group_id\"),\n",
    "        variable.name = \"chr_group\",\n",
    "        value.name =\"gene\")\n",
    "    \n",
    "    list(canonicalTranscripts=canonicalTranscripts, \n",
    "       meanTpms=meanTpms,\n",
    "       triads=triads, \n",
    "       triadMovement=triadMovement,\n",
    "       gene_universe=gene_universe, \n",
    "       ontologies=ontologies,\n",
    "       id_names=id_names,\n",
    "       WGCNA=WGCNA,\n",
    "       GOSlim=go_slim,\n",
    "       partition=partition,\n",
    "       motifs=motifs,\n",
    "       allTriads=allTriads\n",
    "       )\n",
    "}\n",
    "\n",
    "\n",
    "folder<-\"./TablesForExploration\"\n",
    "\n",
    "\n",
    "geneInformation<-loadGeneInformation(dir=folder,motifs=F,meanTpms=T)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>value</th><th scope=col>factor</th><th scope=col>gene</th><th scope=col>samples</th><th scope=col>subset</th><th scope=col>min_mean_tpm</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>  40.6321681        </td><td>all                 </td><td>TraesCS1A01G000100  </td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "\t<tr><td>1247.8227773        </td><td>all                 </td><td>TraesCS1A01G000100LC</td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "\t<tr><td> 357.8591670        </td><td>all                 </td><td>TraesCS1A01G000200  </td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "\t<tr><td>   0.1167718        </td><td>all                 </td><td>TraesCS1A01G000200LC</td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "\t<tr><td>  17.7845964        </td><td>all                 </td><td>TraesCS1A01G000300  </td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "\t<tr><td>   0.3985324        </td><td>all                 </td><td>TraesCS1A01G000300LC</td><td>140                 </td><td>stress              </td><td>0.5                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllll}\n",
       " value & factor & gene & samples & subset & min\\_mean\\_tpm\\\\\n",
       "\\hline\n",
       "\t   40.6321681         & all                  & TraesCS1A01G000100   & 140                  & stress               & 0.5                 \\\\\n",
       "\t 1247.8227773         & all                  & TraesCS1A01G000100LC & 140                  & stress               & 0.5                 \\\\\n",
       "\t  357.8591670         & all                  & TraesCS1A01G000200   & 140                  & stress               & 0.5                 \\\\\n",
       "\t    0.1167718         & all                  & TraesCS1A01G000200LC & 140                  & stress               & 0.5                 \\\\\n",
       "\t   17.7845964         & all                  & TraesCS1A01G000300   & 140                  & stress               & 0.5                 \\\\\n",
       "\t    0.3985324         & all                  & TraesCS1A01G000300LC & 140                  & stress               & 0.5                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "value | factor | gene | samples | subset | min_mean_tpm | \n",
       "|---|---|---|---|---|---|\n",
       "|   40.6321681         | all                  | TraesCS1A01G000100   | 140                  | stress               | 0.5                  | \n",
       "| 1247.8227773         | all                  | TraesCS1A01G000100LC | 140                  | stress               | 0.5                  | \n",
       "|  357.8591670         | all                  | TraesCS1A01G000200   | 140                  | stress               | 0.5                  | \n",
       "|    0.1167718         | all                  | TraesCS1A01G000200LC | 140                  | stress               | 0.5                  | \n",
       "|   17.7845964         | all                  | TraesCS1A01G000300   | 140                  | stress               | 0.5                  | \n",
       "|    0.3985324         | all                  | TraesCS1A01G000300LC | 140                  | stress               | 0.5                  | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  value        factor gene                 samples subset min_mean_tpm\n",
       "1   40.6321681 all    TraesCS1A01G000100   140     stress 0.5         \n",
       "2 1247.8227773 all    TraesCS1A01G000100LC 140     stress 0.5         \n",
       "3  357.8591670 all    TraesCS1A01G000200   140     stress 0.5         \n",
       "4    0.1167718 all    TraesCS1A01G000200LC 140     stress 0.5         \n",
       "5   17.7845964 all    TraesCS1A01G000300   140     stress 0.5         \n",
       "6    0.3985324 all    TraesCS1A01G000300LC 140     stress 0.5         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(geneInformation$meanTpms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.5",
   "language": "R",
   "name": "ir35"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
