# WheatHomoeologExpression


## Methods for Transcriptome paper. 

This repository contains the code as used for the following sections

### Mapping and reference

Philippa, but Ricardo could add the scripts that he used to re-run. 

Filter: Ricardo

### Homoeolog specific mapping by kallisto in nulli-tetrasomic lines

Philippa

### Theoretical read mapping accuracy between homoeologs using SNP distributions

Ricardo? 

### Genome of origin effect on gene expression
Philippa 

### Expression complexity
Ricardo 

### Expressed genes within the Azhurnaya developmental time course (209 samples) and Chinese Spring no stress (123 samples).
    
Details in Jupyter [notebook](01.%20Prepare%20TPM%20table.ipynb)  


### Differential expression (Azhurnaya time course)

Philippa

### eFP browser 

The browser is available [here](http://bar.utoronto.ca/efp_wheat/cgi-bin/efpWeb.cgi)

### Inference of gene families and homoeologous groups

Fred?



2. Calculate triad category
    *  Relative expression levels of the A, B, and D genome homoeologs across triads
    *  Definition of homoeolog expression bias categories
3.  Classify genes by movement
    * Variation in homoeolog expression bias across tissues (Stable/Dynamic triads)
4. Synthetic wheat calculate TPMs.ipynb
    * Analysis of effect on polyploidy on homoeolog expression bias

## Links to external scripts

    