# WheatHomoeologExpression


## Methods for Transcriptome paper. 

This repository contains the code as used for the following sections

### Mapping and reference

Philippa, but Ricardo could add the scripts that he used to re-run. 

Filter: Ricardo

### Homoeolog specific mapping by kallisto in nulli-tetrasomic lines

Philippa

### Theoretical read mapping accuracy between homoeologs using SNP distributions

Ricardo? 

### Genome of origin effect on gene expression
Philippa 

### Expression complexity
Ricardo 

### Expressed genes within the Azhurnaya developmental time course (209 samples) and Chinese Spring no stress (123 samples).
    
Details in Jupyter [notebook](01.%20Prepare%20TPM%20table.ipynb)  


### Differential expression (Azhurnaya time course)

Philippa

### eFP browser 

The browser is available [here](http://bar.utoronto.ca/efp_wheat/cgi-bin/efpWeb.cgi)

### Inference of gene families and homoeologous groups

Fred?


### Independent measures of D genome homoeolog expression bias
Philippa

#### Inference of gene expression level and breadth categories and sub-genome expression bias

#### Genome-wide comparison of expression levels using principal component analysis (PCA)
 
#### Comparison of homoeolog expression levels using one-way analysis of variance (ANOVA)

#### Assessment of family-wise gene expression bias using phylogenetic comparative ANOVA

### Relative expression levels of the A, B, and D genome homoeologs across triads

Details in Jupyter [notebook](02.%20Calculate%20triad%20category.ipynb)  


### Definition of homoeolog expression bias categories

Details in Jupyter [notebook](02.%20Calculate%20triad%20category.ipynb#Definition-of-homoeolog-expression-bias-categories)  

### Analysis of effect on polyploidy on homoeolog expression bias
Details in Jupyter [notebook](02.%20Classify%20genes%20by%20movement) 

### DNA methylation data analysis 
Curtis et al?

### Comparison of RNA-Seq sample classification with DNA methylation

Ricardo and Curtis et al?

### Histone data analysis
Curtis et al? 

### Variation in homoeolog expression bias across tissues (Stable/Dynamic triads).
Details in Jupyter [notebook](04.%20Synthetic%20wheat%20calculate%20TPMs.ipynb)

### Gene Ontology and Plant Ontology term enrichment
Ricardo

### CDS, protein and promoter analysis for triads
The scripts used for this section are in various repositories: 

  * [Extract promoter regions](https://github.com/lucventurini/mikado/blob/f47aa63/util/extract_promoter_regions.py)
  * [Pairwise alignment of triads](https://github.com/TGAC/bioruby-polyploid-tools/blob/master/bin/blast_triads.rb)

### Transposable Element (TE) modeling using CLARITE
Clarite is available [here](https://github.com/jdaron/CLARI-TE)

### TE presence in gene promoters
Sophie

### Enrichment of TE families in gene promoters
Jemima

### TE density in gene promoters
Jemima

### Transcription factor binding site (TFBS) identification
Jemima

### Ka/Ks analysis
 * [MAFFT alignment](https://github.com/TGAC/bioruby-polyploid- tools/blob/master/bin/mafft_triads.rb.)
 * Test (Sophie)


### WGCNA Network construction 
Philippa

### Defining same, similar and divergent expression patterns of triads
Philippa


### Module overlaps
Philippa


### Gene tree for MADS_II in root module 61

### Correlation to stress status

### Genie3
* [Results from Genie3](https://doi.ipk-gatersleben.de/DOI/53148abd-26a1-4ede-802b-c2635af6a725/0dd8224a-34fc-4e3b-8ab8-883d07e52bd2/2/1847940088)

### Identifying high connected hub genes 
Andrea? 

